{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "# path hack for relative import in jupyter notebook\n",
    "import os\n",
    "import sys\n",
    "import time\n",
    "\n",
    "# LIBRARY GLOBAL MODS\n",
    "CELLTYPES = os.path.dirname(os.path.abspath(''))\n",
    "sys.path.append(CELLTYPES)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.datasets import load_digits\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import PCA\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import umap\n",
    "import time\n",
    "%matplotlib inline\n",
    "\n",
    "sns.set(style='white', context='notebook', rc={'figure.figsize':(14,10)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.colors import ListedColormap, BoundaryNorm, NoNorm, Normalize\n",
    "from matplotlib.colorbar import ColorbarBase\n",
    "from matplotlib.cm import get_cmap, ScalarMappable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# need to set cmaps to be used before importing multicell_replot \n",
    "#   as it uses proplot which modifies matplotlib\n",
    "\n",
    "CMAP_SPECTRALR = get_cmap('Spectral_r')\n",
    "GLOBAL_DPI = 450"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Appended to sys path I:\\Development\\Repositories\\biomodels\\celltypes\n"
     ]
    }
   ],
   "source": [
    "from utils.file_io import RUNS_FOLDER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "NOTEBOOK_OUTDIR = RUNS_FOLDER + os.sep + 'explore' + os.sep + 'nb_alignUMAP'\n",
    "os.makedirs(NOTEBOOK_OUTDIR, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TODO: Import reloading while retaining variables in memory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import importlib\n",
    "#importlib.reload(multicell.multicell_replot)\n",
    "\n",
    "#import inspect\n",
    "#print(inspect.getsource(multicell.multicell_replot.replot_scatter_dots))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions: Non-Plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_X_multi(manyruns_paths, subsample_ndim=None, subsample_ensemble=None, smod_last=True, use_01=True):\n",
    "    # any asserts and simple variable settings\n",
    "    if smod_last:\n",
    "        smod = '_last' # oldstyle\n",
    "    else:\n",
    "        smod = ''      # newstyle\n",
    "    \n",
    "    # get data dimensionality and set system size variables\n",
    "    ref_fpath_state = manyruns_paths[0] + os.sep + 'aggregate' + os.sep + 'X_aggregate%s.npz' % smod\n",
    "    ref_X = np.load(ref_fpath_state)['arr_0'].T  # umap wants transpose\n",
    "    npts, ndim = ref_X.shape\n",
    "    print(\"npts, ndim\", npts, ndim)\n",
    "    \n",
    "    if subsample_ensemble is not None:\n",
    "        assert 0 < subsample_ensemble <= npts\n",
    "        nn = subsample_ensemble\n",
    "    else: \n",
    "        nn = npts\n",
    "        \n",
    "    if subsample_ndim is not None:\n",
    "        assert 0 < subsample_ndim <= ndim\n",
    "        kk = subsample_ndim\n",
    "    else:\n",
    "        kk = ndim\n",
    "    \n",
    "    # Step 1: generate X_multi\n",
    "    X_multi = np.zeros((len(manyruns_paths), nn, kk), dtype=int)\n",
    "    for j, manyruns_path in enumerate(manyruns_paths):\n",
    "\n",
    "        #gamma_val = gamma_list[j]\n",
    "\n",
    "        agg_dir = manyruns_path + os.sep + 'aggregate'\n",
    "        fpath_state = agg_dir + os.sep + 'X_aggregate%s.npz' % smod\n",
    "        fpath_energy = agg_dir + os.sep + 'X_energy%s.npz' % smod\n",
    "        fpath_pickle = manyruns_path + os.sep + 'multicell_template.pkl'\n",
    "\n",
    "        X = np.load(fpath_state)['arr_0'].T  # umap wants transpose\n",
    "        X_energies = np.load(fpath_energy)['arr_0'].T  # umap wants transpose (?)\n",
    "        with open(fpath_pickle, 'rb') as pickle_file:\n",
    "            multicell_template = pickle.load(pickle_file)  # unpickling multicell object\n",
    "\n",
    "        #print('accessing', j, manyruns_path)\n",
    "        X_multi[j, :, :] = X[0:nn, 0:kk]\n",
    "    \n",
    "    if use_01:\n",
    "        X_multi = (1 + X_multi) / 2.0\n",
    "        X_multi = X_multi.astype(int)\n",
    "        \n",
    "    num_cells = multicell_template.num_cells\n",
    "    num_genes = multicell_template.num_genes\n",
    "    return X_multi, num_cells, num_genes\n",
    "\n",
    "\n",
    "def gather_data_then_alignedUMAP(\n",
    "    manyruns_paths, manyruns_dirnames, umap_kwargs, \n",
    "    skip_alignment=False, subsample_ndim=None, subsample_ensemble=None, \n",
    "    smod_last=True, use_01=True):\n",
    "    \"\"\"\n",
    "    All of these arguments are prepared in a dictionary \n",
    "        settings_aligned_umap_fig5_v1\n",
    "        settings_aligned_umap_fig5_v2\n",
    "        etc.\n",
    "    and chosen later in the notebook settings_alignment_chosen\n",
    "    \"\"\"\n",
    "    # TODO once this fn done, need to propogate data_subdict object through nb\n",
    "    #  - it should replace aligned_mapper\n",
    "    #  - should also integrate it with alignment settings dict somehow\n",
    "    # TODO should have class or validation function for data_subdict init\n",
    "    # TODO integrate side data like num unqieu states (analagous to energy)\n",
    "    \n",
    "    if smod_last:\n",
    "        smod = '_last'\n",
    "    else:\n",
    "        smod = ''\n",
    "\n",
    "    def build_data_subdict_from_alignedUMAP(data_subdict, X, embedding):\n",
    "        \"\"\"\n",
    "        data_subdict: partially complete data_subdict (path and label keys)\n",
    "        X:            raw data 2d arr (num_runs x num_dimensions)\n",
    "        embedding:    embedded data (num_runs x num_lowdim)\n",
    "        \"\"\"\n",
    "        # Reload data_subdict source files\n",
    "        #  note: we transpose raw data, energies for umap/other functions\n",
    "        manyruns_path = data_subdict['path']\n",
    "        agg_dir = manyruns_path + os.sep + 'aggregate'\n",
    "        fpath_state = agg_dir + os.sep + 'X_aggregate%s.npz' % smod\n",
    "        fpath_energy = agg_dir + os.sep + 'X_energy%s.npz' % smod\n",
    "        fpath_pickle = manyruns_path + os.sep + 'multicell_template.pkl'\n",
    "        X = np.load(fpath_state)['arr_0'].T          \n",
    "        X_energies = np.load(fpath_energy)['arr_0'].T\n",
    "        with open(fpath_pickle, 'rb') as pickle_file:\n",
    "            # unpickling multicell object\n",
    "            multicell_template = pickle.load(pickle_file)  \n",
    "        \n",
    "        # Fill out the rest of data_subdict keys        \n",
    "        num_runs, total_spins = X.shape\n",
    "        data_subdict['data'] = X\n",
    "        data_subdict['index'] = list(range(num_runs))\n",
    "        data_subdict['energies'] = X_energies\n",
    "        data_subdict['num_runs'] = num_runs\n",
    "        data_subdict['total_spins'] = total_spins\n",
    "        data_subdict['multicell_template'] = multicell_template\n",
    "        data_subdict['algos'] = {\n",
    "            'umap': \n",
    "                {'reducer': None,\n",
    "                 'embedding': embedding}\n",
    "        }\n",
    "\n",
    "        # Add extra analysis data used within the notebook\n",
    "        X_nunique = get_unique_states_over_ensemble(\n",
    "            X, \n",
    "            multicell_template.num_cells, \n",
    "            multicell_template.num_genes\n",
    "        )\n",
    "        data_subdict['nunique'] = X_nunique.T  # note transpose\n",
    "        assert (\n",
    "            data_subdict['energies'].shape[0] == \n",
    "            data_subdict['nunique'].shape[0])\n",
    "        \n",
    "        return data_subdict\n",
    "    \n",
    "    # Step 0: any asserts and simple variable settings\n",
    "    assert 'random_state' in umap_kwargs.keys()\n",
    "    \n",
    "    # Step 1: construct X_multi (tensor of tissue state ensembles)\n",
    "    X_multi, _, _ = build_X_multi(\n",
    "        manyruns_paths, \n",
    "        subsample_ndim=subsample_ndim, \n",
    "        subsample_ensemble=subsample_ensemble, \n",
    "        smod_last=smod_last, \n",
    "        use_01=use_01\n",
    "    )\n",
    "    nn = X_multi.shape[1]\n",
    "    \n",
    "    # Step 2: perform alignedUMAP\n",
    "    aligned_mapper = umap.AlignedUMAP(**umap_kwargs)\n",
    "    \n",
    "    if skip_alignment:\n",
    "        print('Warning - skipping alignment step...')\n",
    "        mappers = [\n",
    "            umap.UMAP(**umap_kwargs).fit(X_multi[n, :, :])\n",
    "            for n in range(len(manyruns_paths))\n",
    "        ]\n",
    "        embeddings = [mappers[n].embedding_ for n in range(len(manyruns_paths))]\n",
    "        aligned_mapper.embeddings_ = embeddings\n",
    "    else:\n",
    "        # UMAP aligned needs a 'relation dict' for the 'time varying' dataset\n",
    "        #   our relation is that each traj maps to itself (in time) -- constant relation\n",
    "        constant_dict = {i: i for i in range(nn)}\n",
    "        constant_relations = [constant_dict for i in range(len(manyruns_paths)-1)]\n",
    "        print('Beginning AlignedUMAP...')\n",
    "        aligned_mapper.fit(X_multi, relations=constant_relations)\n",
    "        print('...done')\n",
    "        \n",
    "    # Step 3: Convert to data_subdict format\n",
    "    # TODO - implement saving\n",
    "    embedded_datasets = {i: {'label': manyruns_dirnames[i],\n",
    "                             'path': manyruns_paths[i]}\n",
    "                         for i in range(len(manyruns_dirnames))}\n",
    "    \n",
    "    for idx in range(len(manyruns_dirnames)):\n",
    "        embedded_datasets[idx] = build_data_subdict_from_alignedUMAP(\n",
    "            embedded_datasets[idx], \n",
    "            X_multi[idx,:,:], \n",
    "            aligned_mapper.embeddings_[idx]\n",
    "        )\n",
    "        #fpath = manyruns_paths[idx] + os.sep + 'dimreduce' \\\n",
    "        #        + os.sep + 'dimreduce%s.z' % fmod\n",
    "        #save_dimreduce_object(datasets[idx], fpath)  # save to file (joblib)\n",
    "    \n",
    "    return embedded_datasets\n",
    "\n",
    "\n",
    "def tissue_to_num_unique_cells(X, num_cells, num_genes):\n",
    "    if X.shape != (num_genes, num_cells):\n",
    "        # perform the reshape\n",
    "        assert len(X) == num_genes * num_cells\n",
    "        X = X.reshape((num_genes, num_cells), order='F')\n",
    "        \n",
    "    tissue_unique_cols = np.unique(X, axis=1)\n",
    "    num_unique = tissue_unique_cols.shape[1]\n",
    "    return num_unique, tissue_unique_cols\n",
    "\n",
    "\n",
    "def get_unique_states_over_ensemble(ensemble_array, num_cells, num_genes):\n",
    "    # TODO track with X_energies somehow...\n",
    "    num_ensemble = ensemble_array.shape[0]\n",
    "    carray = np.zeros(num_ensemble, dtype=int)\n",
    "    for idx in range(num_ensemble):\n",
    "        tissue_arr = ensemble_array[idx, :]\n",
    "        num_unique, unique_states = tissue_to_num_unique_cells(tissue_arr, num_cells, num_genes)\n",
    "        carray[idx] = num_unique\n",
    "    return carray"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_PCA_to_embedded_datasets(embedded_datasets):\n",
    "    for key in embedded_datasets.keys():\n",
    "        data_subdict = embedded_datasets[key]\n",
    "        assert 'pca' not in data_subdict['algos'].keys()\n",
    "        data_subdict['algos']['pca'] = {}\n",
    "        # get full PCA embedding\n",
    "        pca_full = PCA()\n",
    "        embedding = pca_full.fit_transform(data_subdict['data'])\n",
    "        data_subdict['algos']['pca']['embedding'] = embedding[:,:2]\n",
    "        # also store cum. sum. pca variance\n",
    "        exp_var_cumul = np.cumsum(pca_full.explained_variance_ratio_)\n",
    "        embedding = data_subdict['algos']['pca']['cumsum_x'] = \\\n",
    "            np.arange(1, exp_var_cumul.shape[0] + 1)\n",
    "        embedding = data_subdict['algos']['pca']['cumsum_y'] = \\\n",
    "            exp_var_cumul\n",
    "    return embedded_datasets\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions: Plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def axis_bounds(embedding):\n",
    "    left, right = embedding.T[0].min(), embedding.T[0].max()\n",
    "    bottom, top = embedding.T[1].min(), embedding.T[1].max()\n",
    "    adj_h, adj_v = (right - left) * 0.1, (top - bottom) * 0.1\n",
    "    return [left - adj_h, right + adj_h, bottom - adj_v, top + adj_v]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_fpdata_as_grid(\n",
    "    embedded_datasets, settings_alignment, fpdata_type, fpath, \n",
    "    gamma_selection=None, nrow=2, ncol=4, ext='.jpg', logy=False): \n",
    "    \"\"\"\n",
    "    Fixed point (FP) data forms: \n",
    "    - 'energies': energy array (1 dimensional -- 1 dim per fp)\n",
    "    - 'nunique': number of unique single cell states (1 dimensional -- 1 dim per fp)\n",
    "    Args:\n",
    "    - embedded_datasets: dict of data_subdicts objects\n",
    "    - settings_alignment: dict of manyruns info from jupyter notebook globals\n",
    "    - fpdata_type: 'energy' or 'nunique'\n",
    "    - gamma_selection: list of integers to choose a subset of data from settings_alignment['manyruns_paths']\n",
    "    \"\"\"\n",
    "    \n",
    "    gamma_values = settings_alignment['gamma_list']\n",
    "    manyruns_paths = settings_alignment['manyruns_paths']\n",
    "    #nn = settings_alignment['alignment_wrapper_kwargs']['subsample_ensemble']\n",
    "    if gamma_selection is None:\n",
    "        gamma_selection = settings_alignment['subplot_selection']\n",
    "    \n",
    "    assert fpdata_type in ['energies', 'nunique']\n",
    "    # TODO smarter bin settings: set after load data, observe statistics?\n",
    "    # TODO bins_settings = np.arange(0.5,40,1)  \n",
    "    data_arrays = [0] * len(gamma_selection)\n",
    "    if fpdata_type == 'energies':\n",
    "        bins_settings = 40\n",
    "        color = 'coral'\n",
    "        xlabel = 'Energy'\n",
    "        for idx, idx_selected in enumerate(gamma_selection):\n",
    "            data_arrays[idx] = embedded_datasets[idx_selected][fpdata_type][:, 0]\n",
    "    else:\n",
    "        bins_settings = np.arange(0.5,40,1)  # TODO identify after loading data and observing statistics? \n",
    "        color = 'deepskyblue'\n",
    "        xlabel = 'Number of unique cell states'\n",
    "        for idx, idx_selected in enumerate(gamma_selection):\n",
    "            data_arrays[idx] = embedded_datasets[idx_selected][fpdata_type]\n",
    "    \n",
    "    # plot settings\n",
    "    if logy:\n",
    "        fpath += '_logy'\n",
    "        \n",
    "    # prints\n",
    "    print('Picks for plot_fpdata_as_grid():')\n",
    "    print([(i, gamma_values[i]) for i in gamma_selection])\n",
    "\n",
    "    # Step 2: plot as grid of histograms\n",
    "    fig, axs = plt.subplots(nrow, ncol, figsize=(18*0.68, 12*0.68))\n",
    "    for idx, ax in enumerate(axs.flatten()):\n",
    "        gamma_index = gamma_selection[idx]\n",
    "        gamma_value = gamma_values[gamma_index]\n",
    "        ptitle = r'%d: idx=%d, $\\gamma=%.3f$' % (\n",
    "            idx, gamma_index, gamma_value)\n",
    "        ax.set_title(ptitle, fontsize=12)\n",
    "        ax.hist(data_arrays[idx], bins=bins_settings, color=color)\n",
    "        if logy:\n",
    "            ax.set_yscale('log')\n",
    "    \n",
    "    # Step 3: global axis labels\n",
    "    offset = 0.005\n",
    "    fig.text(0.5, offset, xlabel, ha='center')\n",
    "    fig.text(offset, 0.5, 'Number of fixed points', \n",
    "             va='center', rotation='vertical')\n",
    "\n",
    "    #ax.axis(ax_bound)\n",
    "    #ax.set(xticks=[], yticks=[])\n",
    "            \n",
    "    plt.tight_layout(pad=1.2, w_pad=0.5, h_pad=1.0)\n",
    "    if isinstance(ext, list):\n",
    "        for ext_str in ext:\n",
    "            assert ext_str[0] == '.'\n",
    "            plt.savefig(fpath + ext_str, dpi=300)\n",
    "    else: \n",
    "        plt.savefig(fpath + ext, dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_a_point(ax, embedding, agg_idx, ss=10, zorder=4):\n",
    "    x, y = embedding[agg_idx, :]\n",
    "    ax.scatter(x, y, s=ss, facecolors='none', edgecolors='k',\n",
    "              path_effects=[PathEffects.withStroke(linewidth=2, foreground=\"w\")],\n",
    "              zorder=zorder)\n",
    "    ax.annotate('%d' % agg_idx, (x, y), \n",
    "                xytext=(3, 3),\n",
    "                textcoords='offset points',\n",
    "                path_effects=[PathEffects.withStroke(linewidth=2, foreground=\"w\")],\n",
    "                zorder=zorder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.patheffects as PathEffects\n",
    "\n",
    "def plot_aligned_panels_as_grid(\n",
    "    embedded_datasets, settings_alignment, fpath, \n",
    "    algo='umap',\n",
    "    gamma_selection=None, ext='.jpg', \n",
    "    ss=6, annotate_indices=[], rasterized=True,\n",
    "    c='energies', skip_cbar=False,\n",
    "    alpha=1.0, edges=False, black_edges=True):\n",
    "    \"\"\"\n",
    "    Args: \n",
    "    - embedded_datasets: dict of data_subdicts (stores metadata, embedding)\n",
    "    - settings_alignment: dict of settings for the notebook,\n",
    "    - fpath: output path for plots (minus extension)\n",
    "    - algo: (str) key for data_subdict['algos'][algo] (e.g. 'umap', 'pca')\n",
    "    - gamma_selection: can override same key in settings_alignment\n",
    "    - ss: change marker size for scatter\n",
    "    - annotate_indices: list of indices to annotate on the embedding\n",
    "    - rasterized: If True, rasterize the scatter points at given dpi\n",
    "    - edges: if True, plot edges\n",
    "        - markers and edges separate calls to scatter\n",
    "        - alpha affects marker alpha only\n",
    "        - edge color default: face color (try switch to black)\n",
    "    Settings to consider changing:\n",
    "    - fontsize for title\n",
    "    Good defaults:\n",
    "    - ss=6, edges=True (with linewidth 0.1)\n",
    "    Notes:\n",
    "    - nunique cmap is based on \n",
    "        sns.color_palette(\"pastel\") or\n",
    "        sns.color_palette()\n",
    "        see https://seaborn.pydata.org/generated/seaborn.color_palette.html\n",
    "    \n",
    "    \"\"\"\n",
    "    # this line maintains cmap colors (matplotlib and proplot fighting)\n",
    "    sns.set(style='white', context='notebook')\n",
    "    dpi = GLOBAL_DPI\n",
    "    \n",
    "    # parse info from dict: settings_alignment\n",
    "    gamma_list = settings_alignment['gamma_list']\n",
    "    manyruns_paths = settings_alignment['manyruns_paths']\n",
    "    nn = settings_alignment['alignment_wrapper_kwargs']['subsample_ensemble']\n",
    "    assert nn is None\n",
    "    if gamma_selection is None:\n",
    "        gamma_selection = settings_alignment['subplot_selection']\n",
    "    \n",
    "    # parse info from dict: embedded_datasets (contains data_subdicts)\n",
    "    # TODO handler for embedding dimension > 2?\n",
    "    assert algo in embedded_datasets[0]['algos'].keys()\n",
    "    embeddings = [\n",
    "        embedded_datasets[idx]['algos'][algo]['embedding'][:, 0:2]\n",
    "        for idx in range(len(settings_alignment_chosen['gamma_list']))\n",
    "    ]\n",
    "\n",
    "    # Part A: Figure initialization\n",
    "    # see https://matplotlib.org/stable/tutorials/intermediate/gridspec.html\n",
    "    num_rows = 2\n",
    "    num_cols = 4\n",
    "    if skip_cbar:\n",
    "        #fig = plt.figure(figsize=(5,5), dpi=dpi)\n",
    "        fig, axs = plt.subplots(\n",
    "            num_rows, num_cols, figsize=(12.24, 8.16), dpi=dpi)\n",
    "    else:\n",
    "        cbar_horizontal = False\n",
    "        if cbar_horizontal:\n",
    "            fig, axs = plt.subplots(\n",
    "                (num_rows+1), num_cols, figsize=(12.24, 8.16), dpi=dpi,\n",
    "                gridspec_kw={'height_ratios': [1,1,0.1]}\n",
    "            )\n",
    "            orientation = 'horizontal'\n",
    "            gs = axs[-1,-1].get_gridspec()\n",
    "            cbar_ax = fig.add_subplot(gs[-1, :])\n",
    "            for ax in axs[-1, :]:\n",
    "                ax.remove()\n",
    "        else:\n",
    "            fig, axs = plt.subplots(\n",
    "                num_rows, (num_cols+1), figsize=(12.24, 8.16), dpi=dpi, \n",
    "                gridspec_kw={'width_ratios': [1,1,1,1,0.1]}\n",
    "            )\n",
    "            orientation = 'vertical'\n",
    "            gs = axs[0, -1].get_gridspec()\n",
    "            cbar_ax = fig.add_subplot(gs[:, -1])\n",
    "            for ax in axs[:, -1]:\n",
    "                ax.remove()\n",
    "        cbar_ax.set(xticks=[], yticks=[])\n",
    "        \n",
    "    # Part B: Plot coloring rules (static, nunique, or energy)\n",
    "    if c is None:\n",
    "        fixed_color = 'gainsboro'  # '#5076b7', 'slategray', 'gainsboro'\n",
    "        cmap = None\n",
    "    else:\n",
    "        assert c in ['energies', 'nunique']\n",
    "        if c == 'energies':\n",
    "            #cmap_str = 'Spectral_r'\n",
    "            #cmap = get_cmap(cmap_str)\n",
    "            cmap = CMAP_SPECTRALR\n",
    "            \n",
    "            if not skip_cbar:\n",
    "                artificial_norm = Normalize(vmin=-1, vmax=1)\n",
    "                mappable = ScalarMappable(artificial_norm, cmap=cmap)\n",
    "                cbar = plt.colorbar(\n",
    "                    mappable,\n",
    "                    cax=cbar_ax,\n",
    "                    orientation=orientation,\n",
    "                )\n",
    "                cbar.set_label('Energy')\n",
    "                cbar.ax.tick_params(size=0)\n",
    "                if cbar_horizontal:\n",
    "                    cbar.ax.set_xticklabels([])\n",
    "                else:\n",
    "                    cbar.ax.set_yticklabels([])\n",
    "                #cbar_ax.set_xticks([])\n",
    "                #cbar_ax.set(xticks=[], yticks=[])\n",
    "            \n",
    "        else:\n",
    "            # create bounds for custom discrete cmap\n",
    "            #nunique = data_subdict['nunique']\n",
    "            bounds_upper = 22 #  np.max(nunique), or static if proportional cbar\n",
    "            bounds = [1, 2, 3, 4, 8, 15, bounds_upper]\n",
    "            bounds_midpts = [0.5 * (bounds[i] + bounds[i+1]) for i in range(len(bounds) - 1)]\n",
    "            ncolors = len(bounds) - 1\n",
    "            # refining bounds labels\n",
    "            bounds_labels = [''] * ncolors\n",
    "            for i in range(len(bounds) - 2):\n",
    "                lower = bounds[i]\n",
    "                upper = bounds[i+1] - 1\n",
    "                if lower == upper:\n",
    "                    bounds_labels[i] = '%d' % lower\n",
    "                else:\n",
    "                    bounds_labels[i] = '%d to %d' % (lower, upper)\n",
    "            #bounds_labels = ['%d to %d' % (bounds[i], bounds[i+1] - 1) for i in range(len(bounds) - 1)]\n",
    "            bounds_labels[-1] = r'$\\geq%d$' % bounds[-2]\n",
    "\n",
    "            spectral_auto = False\n",
    "            if spectral_auto:\n",
    "                #cmap_primitive = get_cmap('Spectral_r')\n",
    "                cmap_primitive = CMAP_SPECTRALR\n",
    "                norm_manual = (np.array(bounds_midpts) - bounds_midpts[0]) / (bounds_midpts[-1] - bounds_midpts[0])\n",
    "                print(norm_manual)\n",
    "                color_palette = [cmap_primitive(a) for a in np.linspace(0, 1, ncolors)]\n",
    "            else:\n",
    "                sns_pastel_indices = [0,9,2,3,4,8]\n",
    "                \n",
    "                #color_palette = [sns.color_palette(\"pastel\")[i] \n",
    "                #                 for i in sns_pastel_indices] \n",
    "                color_palette = [sns.color_palette()[i] \n",
    "                                 for i in sns_pastel_indices] \n",
    "                assert len(color_palette) >= ncolors\n",
    "                color_palette = color_palette[0:ncolors]\n",
    "            cmap = ListedColormap(color_palette)\n",
    "            \n",
    "            # create scatter array of colors\n",
    "            norm_nunique = BoundaryNorm(bounds, cmap.N, clip=True)\n",
    "            #color = cmap(norm(nunique))\n",
    "            \n",
    "            if not skip_cbar:\n",
    "                # this is alternative to gridspec:\n",
    "                #fig.subplots_adjust(right=0.8)\n",
    "                #cbar_ax = fig.add_axes([0.85, 0.15, 0.05, 0.7])\n",
    "                cbar = ColorbarBase(\n",
    "                    cbar_ax,\n",
    "                    cmap=cmap,\n",
    "                    norm=norm_nunique,\n",
    "                    #boundaries=[0] + bounds + [13],\n",
    "                    ticks=bounds_midpts,\n",
    "                    spacing='proportional',  # 'proportional', 'uniform'\n",
    "                    orientation=orientation)\n",
    "                if cbar_horizontal:\n",
    "                    cbar.ax.set_xticklabels(bounds_labels)\n",
    "                else:\n",
    "                    cbar.ax.set_yticklabels(bounds_labels)\n",
    "                cbar.set_label('Number of unique single cell states')\n",
    "    \n",
    "    # Part C: Edges (on scatterplot points) or not? \n",
    "    sc_kwargs = { \n",
    "        'cmap': cmap,\n",
    "        's': ss,\n",
    "        'rasterized': rasterized,\n",
    "        'alpha': alpha}\n",
    "    sc_kwargs['edgecolors'] = 'None'\n",
    "    if edges:\n",
    "        #sc_kwargs['edgecolors'] = 'None'\n",
    "        sc_kwargs_edge = sc_kwargs.copy()\n",
    "        sc_kwargs_edge.pop('edgecolors')\n",
    "        sc_kwargs_edge['alpha'] = 1.0\n",
    "        sc_kwargs_edge['facecolor'] = 'None'\n",
    "        sc_kwargs_edge['linewidths'] = 0.4  # 0.1  # ss / 20.0\n",
    "        \n",
    "    # Part D: Fill in each grid box (default - 2x4 panels)\n",
    "    ax_bound = axis_bounds(np.vstack(embeddings))\n",
    "\n",
    "    #for j, ax in enumerate(axs.flatten()):\n",
    "        #if j < len(gamma_selection):\n",
    "    \n",
    "    for j in range(num_rows * num_cols):\n",
    "        row_idx = j // num_cols\n",
    "        col_idx = j % num_cols\n",
    "        ax = axs[row_idx, col_idx] \n",
    "        i = gamma_selection[j]\n",
    "\n",
    "        data_subdict = embedded_datasets[i]\n",
    "        if c == 'energies':\n",
    "            color_arr = data_subdict['energies'][:, 0]\n",
    "            cnorm = Normalize(\n",
    "                vmin=np.min(color_arr), vmax=np.max(color_arr))\n",
    "            colors = cmap(cnorm(color_arr))\n",
    "        elif c == 'nunique':\n",
    "            color_arr = data_subdict['nunique']\n",
    "            #cnorm = Normalize(\n",
    "            #    vmin=np.min(color_arr), vmax=np.max(color_arr))\n",
    "            colors = cmap(norm_nunique(color_arr))\n",
    "        else:\n",
    "            colors = fixed_color\n",
    "\n",
    "        ax.set_title(gamma_list[i], fontsize=20, y=-0.12)\n",
    "\n",
    "        sc = ax.scatter(*embeddings[i].T, \n",
    "                        **sc_kwargs, \n",
    "                        c=colors,\n",
    "                        zorder=3)\n",
    "        # Use two scatter calls to get edges with different alpha\n",
    "        if edges:\n",
    "            if black_edges:\n",
    "                edgecolors = 'black'\n",
    "            else:\n",
    "                edgecolors = colors\n",
    "            sc2 = ax.scatter(*embeddings[i].T, \n",
    "                             **sc_kwargs_edge, \n",
    "                             edgecolors=edgecolors,\n",
    "                             zorder=2)\n",
    "\n",
    "        # annotate certain elements with empty circles\n",
    "        if isinstance(annotate_indices, dict):\n",
    "            # dict of lists for each gamma index\n",
    "            annotate_indices_specific = annotate_indices[i]\n",
    "            for agg_idx in annotate_indices_specific:\n",
    "                annotate_a_point(ax, embeddings[i], agg_idx, zorder=4)\n",
    "        else:\n",
    "            # assume its the same list for each gamma value\n",
    "            for agg_idx in annotate_indices:\n",
    "                annotate_a_point(ax, embeddings[i], agg_idx, zorder=4)\n",
    "\n",
    "        #ax.axis(ax_bound)\n",
    "        ax.set(xticks=[], yticks=[])\n",
    "\n",
    "    # Part E: Plot Save\n",
    "    plt.tight_layout()\n",
    "    if isinstance(ext, list):\n",
    "        for ext_str in ext:\n",
    "            assert ext_str[0] == '.'\n",
    "            plt.savefig(fpath + ext_str, dpi=dpi)\n",
    "    else: \n",
    "        plt.savefig(fpath + ext, dpi=dpi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_single_umap_panel_gridspec(\n",
    "    data_subdict, fpath, \n",
    "    c=None, ext='.jpg', ss=10, annotate_indices=[], \n",
    "    rasterized=True, skip_cbar=False, \n",
    "    alpha=1.0, edges=False):\n",
    "    # this line maintains cmap colors (matplotlib and proplot fighting)\n",
    "    sns.set(style='white', context='notebook')\n",
    "    dpi = GLOBAL_DPI\n",
    "    \n",
    "    num_runs = data_subdict['num_runs']\n",
    "    label = data_subdict['label']\n",
    "    embedding = data_subdict['algos']['umap']['embedding']\n",
    "    assert embedding.shape[1] == 2\n",
    "    \n",
    "    if skip_cbar:\n",
    "        fig = plt.figure(figsize=(5,5), dpi=dpi)\n",
    "    else:\n",
    "        cbar_horizontal = True\n",
    "        if cbar_horizontal:\n",
    "            fig, ax = plt.subplots(\n",
    "                2, 1, figsize=(5.5, 5), dpi=dpi,\n",
    "                gridspec_kw={'height_ratios': [12, 1]}\n",
    "            )\n",
    "            orientation = 'horizontal'\n",
    "        else:\n",
    "            fig, ax = plt.subplots(\n",
    "                1, 2, figsize=(5, 5.5), dpi=dpi, \n",
    "                gridspec_kw={'width_ratios': [12, 1]}\n",
    "            )\n",
    "            orientation = 'vertical'\n",
    "\n",
    "    if c is None:\n",
    "        color = 'gainsboro'  # '#5076b7', 'slategray', 'gainsboro'\n",
    "        cmap = None\n",
    "    else:\n",
    "        assert c in ['energies', 'nunique']\n",
    "        if c == 'energies':\n",
    "            color = data_subdict['energies'][:, 0]  # range(num_runs)\n",
    "            cmap = 'Spectral_r'\n",
    "        else:\n",
    "            nunique = data_subdict['nunique']\n",
    "            \n",
    "            bounds_upper = 22 #  np.max(nunique), or static if proportional cbar\n",
    "            bounds = [1, 2, 3, 4, 8, 15, bounds_upper]\n",
    "            bounds_midpts = [0.5 * (bounds[i] + bounds[i+1]) for i in range(len(bounds) - 1)]\n",
    "            ncolors = len(bounds) - 1\n",
    "            # refining bounds labels\n",
    "            bounds_labels = [''] * ncolors\n",
    "            for i in range(len(bounds) - 2):\n",
    "                lower = bounds[i]\n",
    "                upper = bounds[i+1] - 1\n",
    "                if lower == upper:\n",
    "                    bounds_labels[i] = '%d' % lower\n",
    "                else:\n",
    "                    bounds_labels[i] = '%d to %d' % (lower, upper)\n",
    "            #bounds_labels = ['%d to %d' % (bounds[i], bounds[i+1] - 1) for i in range(len(bounds) - 1)]\n",
    "            bounds_labels[-1] = r'$\\geq%d$' % bounds[-2]\n",
    "\n",
    "            spectral_auto = False\n",
    "            if spectral_auto:\n",
    "                #cmap_primitive = get_cmap('Spectral_r')\n",
    "                cmap_primitive = CMAP_SPECTRALR\n",
    "                norm_manual = (np.array(bounds_midpts) - bounds_midpts[0]) / (bounds_midpts[-1] - bounds_midpts[0])\n",
    "                print(norm_manual)\n",
    "                color_palette = [cmap_primitive(a) for a in np.linspace(0, 1, ncolors)]\n",
    "            else:\n",
    "                #colors_bank = [cmap_primitive(a) for a in \n",
    "                #               [0.0, 0.1, 0.15, 0.25, 0.35, 0.55, 0.65, 0.75, 0.85, 1.0]]\n",
    "                \n",
    "                # 12 max for sns \"Paired\", manually rearrange the color list first\n",
    "                #sns_paired_indices = [0,1,2,3,9,8]\n",
    "                #sns_paired_indices = [0,1,2,3,7,6]\n",
    "                #color_palette = [sns.color_palette(\"Paired\")[i] \n",
    "                #                 for i in sns_paired_indices] \n",
    "                \n",
    "                #sns_set2_indices = [0,1,2,3,4,7]\n",
    "                #color_palette = [sns.color_palette(\"Set2\")[i] \n",
    "                #                 for i in sns_set2_indices] \n",
    "                \n",
    "                #sns_pastel_indices = [0,1,2,3,4,6]\n",
    "                #sns_pastel_indices = [0,1,2,3,4,9]\n",
    "                #sns_pastel_indices = [0,7,2,3,4,9]\n",
    "                #sns_pastel_indices = [0,9,2,3,4,7]\n",
    "                sns_pastel_indices = [0,9,2,3,4,8]\n",
    "                #color_palette = [sns.color_palette(\"pastel\")[i] \n",
    "                #                 for i in sns_pastel_indices] \n",
    "                color_palette = [sns.color_palette()[i] \n",
    "                                 for i in sns_pastel_indices]\n",
    "                \n",
    "                assert len(color_palette) >= ncolors\n",
    "                color_palette = color_palette[0:ncolors]\n",
    "                \n",
    "                \n",
    "            cmap = ListedColormap(color_palette)\n",
    "            \n",
    "            # create scatter array of colors\n",
    "            norm = BoundaryNorm(bounds, cmap.N, clip=True)\n",
    "            color = cmap(norm(nunique))\n",
    "            \n",
    "            # use grid spec to make two axes for this block\n",
    "            if not skip_cbar:\n",
    "                cbar = ColorbarBase(\n",
    "                    ax[1], \n",
    "                    cmap=cmap,\n",
    "                    norm=norm,\n",
    "                    #boundaries=[0] + bounds + [13],\n",
    "                    ticks=bounds_midpts,\n",
    "                    spacing='proportional',  # 'proportional', 'uniform'\n",
    "                    orientation=orientation)\n",
    "                if cbar_horizontal:\n",
    "                    cbar.ax.set_xticklabels(bounds_labels)\n",
    "                else:\n",
    "                    cbar.ax.set_yticklabels(bounds_labels)\n",
    "                cbar.set_label('Number of unique single cell states')\n",
    "            \n",
    "    sc_kwargs = {\n",
    "        'c': color, \n",
    "        'cmap': cmap,\n",
    "        's': ss,\n",
    "        'rasterized': rasterized,\n",
    "        'alpha': alpha        \n",
    "    }\n",
    "    sc_kwargs['edgecolors'] = 'None'\n",
    "    if edges:\n",
    "        sc_kwargs_edge = sc_kwargs.copy()\n",
    "        sc_kwargs_edge['alpha'] = 1.0\n",
    "        sc_kwargs_edge['facecolor'] = 'None'\n",
    "        sc_kwargs_edge['linewidths'] = 0.25\n",
    "        sc_kwargs_edge['edgecolors'] = 'black'\n",
    "            \n",
    "    if skip_cbar:\n",
    "        ax_sc = plt.gca()\n",
    "    else: \n",
    "        ax_sc = ax[0]\n",
    "    sc = ax_sc.scatter(\n",
    "        embedding[:, 0], embedding[:, 1], **sc_kwargs, zorder=3)\n",
    "    if edges:\n",
    "        sc2 = ax_sc.scatter(\n",
    "            embedding[:, 0], embedding[:, 1], **sc_kwargs_edge, zorder=2)\n",
    "    #ax_sc.set_aspect('equal', 'datalim')\n",
    "    ax_sc.set(xticks=[], yticks=[])\n",
    "    \n",
    "    # annotate certain elements with empty circles\n",
    "    annotate_ss = 15\n",
    "    if isinstance(annotate_indices, dict):\n",
    "        # dict of lists for each gamma index\n",
    "        annotate_indices_specific = annotate_indices[i]\n",
    "        for agg_idx in annotate_indices_specific:\n",
    "            annotate_a_point(ax_sc, embedding, agg_idx, ss=annotate_ss,\n",
    "                             zorder=4)\n",
    "    else:\n",
    "        # assume its the same list for each gamma value\n",
    "        for agg_idx in annotate_indices:\n",
    "            annotate_a_point(ax_sc, embedding, agg_idx, ss=annotate_ss, \n",
    "                             zorder=4)\n",
    "    \n",
    "    if c == 'energies' and not skip_cbar:\n",
    "        cbar = plt.colorbar(sc, cax=ax[1])\n",
    "        cbar.ax.tick_params(size=0)\n",
    "        #plt.colorbar(boundaries=np.arange(11)-0.5).set_ticks(np.arange(10))\n",
    "\n",
    "    # plt.title('UMAP projection of the %s dataset' % label, fontsize=24)\n",
    "    \n",
    "    plt.tight_layout()\n",
    "    if isinstance(ext, list):\n",
    "        for ext_str in ext:\n",
    "            assert ext_str[0] == '.'\n",
    "            plt.savefig(fpath + ext_str, dpi=dpi)\n",
    "    else: \n",
    "        plt.savefig(fpath + ext, dpi=dpi)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Specify settings and gen/load alignedUMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "RANDOM_STATE = 6  #0 #100 #40 #20 #0  Note RANDOM_STATE = 6 for July fig. generation\n",
    "\n",
    "UMAP_KWARGS_DEFAULT = {\n",
    "    'random_state': RANDOM_STATE,  \n",
    "    'transform_seed': RANDOM_STATE,\n",
    "    'n_components': 2,\n",
    "    'metric': 'euclidean',\n",
    "    'init': 'spectral',\n",
    "    'unique': False,\n",
    "    'n_neighbors': 15,  # Default: 15\n",
    "    'min_dist': 0.1,    # Default: 0.1\n",
    "    'spread': 1.0,      # Default: 1.0\n",
    "    'n_epochs': 200,    # Default: 200\n",
    "}\n",
    "\n",
    "UMAP_KWARGS_ALT = {\n",
    "    'random_state': RANDOM_STATE,  \n",
    "    'transform_seed': RANDOM_STATE,\n",
    "    'n_components': 2,\n",
    "    'metric': 'euclidean',\n",
    "    'init': 'spectral',\n",
    "    'unique': False,\n",
    "    'n_neighbors': 15,  # Default: 15\n",
    "    'min_dist': 0.1,    # Default: 0.1\n",
    "    'spread': 1.0,      # Default: 1.0\n",
    "    'n_epochs': 200     # Default: 200\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO - consider also including a subset parameter for relationsdict (subset to 900 was the preJUly Fig5)\n",
    "\n",
    "# Settings used to generate Fig 4, 5, associated SI Figures\n",
    "gamma_list_v1 = [0.0, 0.05, 0.06, 0.07, 0.08, \n",
    "                 0.09, 0.10, 0.15, 0.20, 0.4, \n",
    "                 0.6, 0.8, 0.9, 1.0, 20.0]\n",
    "manyruns_dirnames_v1 = [\n",
    "    'Wrandom0_gamma%.2f_10k_periodic_fixedorderV3_p3_M100' % a \n",
    "    for a in gamma_list_v1]\n",
    "manyruns_paths_v1 = [RUNS_FOLDER + os.sep + 'multicell_manyruns' + os.sep + dirname for dirname in manyruns_dirnames_v1]\n",
    "settings_aligned_umap_fig5_v1 = {\n",
    "    'gamma_list': gamma_list_v1,\n",
    "    'manyruns_paths': manyruns_paths_v1,\n",
    "    'manyruns_dirnames': manyruns_dirnames_v1,\n",
    "    'umap_kwargs': UMAP_KWARGS_DEFAULT,\n",
    "    'alignment_wrapper_kwargs': \n",
    "        {'subsample_ndim': None, \n",
    "         'subsample_ensemble': None,  # 4000 originally\n",
    "         'smod_last': True,\n",
    "         'use_01': False,\n",
    "         'skip_alignment': False},\n",
    "    'subplot_selection': [0, 3, 4, 5, 7, 8, 9, 13]\n",
    "}\n",
    "\n",
    "gamma_list_v2 = [0.0, 0.07, 0.08, 0.09, 0.15, 0.20, 0.4, 1.0]\n",
    "manyruns_dirnames_v2 = [\n",
    "    'Wrandom0_gamma%.2f_10k_periodic_fixedorderV3_p3_M100' % a \n",
    "    for a in gamma_list_v2]\n",
    "manyruns_paths_v2 = [RUNS_FOLDER + os.sep + 'multicell_manyruns' + os.sep + dirname for dirname in manyruns_dirnames_v2]\n",
    "settings_aligned_umap_fig5_v2 = {\n",
    "    'gamma_list': gamma_list_v2,\n",
    "    'manyruns_paths': manyruns_paths_v2,\n",
    "    'manyruns_dirnames': manyruns_dirnames_v2,\n",
    "    'umap_kwargs': UMAP_KWARGS_ALT,\n",
    "    'alignment_wrapper_kwargs': \n",
    "        {'subsample_ndim': None, \n",
    "         'subsample_ensemble': None,\n",
    "         'smod_last': True,\n",
    "         'use_01': False,\n",
    "         'skip_alignment': False},\n",
    "    'subplot_selection': np.arange(8)    \n",
    "}\n",
    "\n",
    "\n",
    "gamma_list_v3 = gamma_list_v1\n",
    "manyruns_dirnames_v3 = [\n",
    "    'Wrandom0_gamma%.2f_10k_periodic_R1_p3_M100' % a \n",
    "    for a in gamma_list_v3]\n",
    "manyruns_paths_v3 = [RUNS_FOLDER + os.sep + 'multicell_manyruns' + os.sep + dirname for dirname in manyruns_dirnames_v3]\n",
    "settings_aligned_umap_fig5_v3 = {\n",
    "    'gamma_list': gamma_list_v3,\n",
    "    'manyruns_paths': manyruns_paths_v3,\n",
    "    'manyruns_dirnames': manyruns_dirnames_v3,\n",
    "    'umap_kwargs': UMAP_KWARGS_DEFAULT,\n",
    "    'alignment_wrapper_kwargs': \n",
    "        {'subsample_ndim': None, \n",
    "         'subsample_ensemble': None,\n",
    "         'smod_last': True,\n",
    "         'use_01': False,\n",
    "         'skip_alignment': False},\n",
    "    'subplot_selection': [0, 3, 4, 5, 7, 8, 9, 13]    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Choose settings option and embed/align UMAP for each manyruns**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "npts, ndim 10000 900\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\users\\matts\\appdata\\local\\programs\\python\\python38\\lib\\site-packages\\proplot\\colors.py:976: ProPlotWarning: Shifting non-cyclic colormap 'acton_truncate'. Use cmap.set_cyclic(True) or Colormap(..., cyclic=True) to suppress this warning.\n",
      "c:\\users\\matts\\appdata\\local\\programs\\python\\python38\\lib\\site-packages\\proplot\\colors.py:976: ProPlotWarning: Shifting non-cyclic colormap 'Sunrise_truncate'. Use cmap.set_cyclic(True) or Colormap(..., cyclic=True) to suppress this warning.\n",
      "c:\\users\\matts\\appdata\\local\\programs\\python\\python38\\lib\\site-packages\\proplot\\colors.py:976: ProPlotWarning: Shifting non-cyclic colormap 'Sunset_truncate'. Use cmap.set_cyclic(True) or Colormap(..., cyclic=True) to suppress this warning.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DYNAMICS_FIXED_UPDATE_ORDER: True\n",
      "Beginning AlignedUMAP...\n"
     ]
    }
   ],
   "source": [
    "flag_generate = True  # If False, load data instead\n",
    "\n",
    "if flag_generate:\n",
    "    settings_alignment_chosen = settings_aligned_umap_fig5_v3\n",
    "    \n",
    "    # these asserts are needed for replotting the high dim states as tissues \n",
    "    #  using raw data within settings_alignment_chosen\n",
    "    assert not settings_alignment_chosen['alignment_wrapper_kwargs']['use_01']\n",
    "    assert settings_alignment_chosen['alignment_wrapper_kwargs']['subsample_ndim'] is None\n",
    "    assert settings_alignment_chosen['alignment_wrapper_kwargs']['subsample_ensemble'] is None\n",
    "\n",
    "    start = time.time()\n",
    "    embedded_datasets = gather_data_then_alignedUMAP(\n",
    "        settings_alignment_chosen['manyruns_paths'], \n",
    "        settings_alignment_chosen['manyruns_dirnames'],\n",
    "        settings_alignment_chosen['umap_kwargs'],\n",
    "        **settings_alignment_chosen['alignment_wrapper_kwargs']\n",
    "    )\n",
    "    end = time.time()\n",
    "    print('Total time (s):', end - start)\n",
    "    \n",
    "    # save embeddings and settings\n",
    "    # TODO - reimplement with new object embedded_datasets\n",
    "    #  - maybe best to save and load each individually? \n",
    "    \"\"\"pickled_alignUMAP = NOTEBOOK_OUTDIR + os.sep + 'aligned_mapper.pkl'\n",
    "    with open(pickled_alignUMAP, 'wb') as f:\n",
    "        # convert from class 'numba.typed.typedlist.List' to numpy array\n",
    "        raw_python_list = [i for i in aligned_mapper.embeddings_]\n",
    "        pickle.dump(raw_python_list, f)\n",
    "    \"\"\"\n",
    "    \n",
    "    # save settings \n",
    "    pickled_align_misc = NOTEBOOK_OUTDIR + os.sep + 'aligned_misc.pkl'\n",
    "    pickle.dump(settings_alignment_chosen, open(pickled_align_misc, 'wb'))\n",
    "    \n",
    "else:\n",
    "    # load settings\n",
    "    input_pickled_align_misc = NOTEBOOK_OUTDIR + os.sep + 'aligned_misc.pkl'\n",
    "    with open(input_pickled_alignUMAP, 'rb') as pickle_file:\n",
    "        settings_alignment_chosen = pickle.load(pickle_file)\n",
    "\n",
    "    # load main object (alignUMAP)\n",
    "    # TODO reimplement as above\n",
    "    \"\"\"\n",
    "    input_pickled_alignUMAP = NOTEBOOK_OUTDIR + os.sep + 'aligned_mapper.pkl'\n",
    "    with open(input_pickled_alignUMAP, 'rb') as pickle_file:\n",
    "        aligned_mapper_embeddings = pickle.load(pickle_file)\n",
    "    aligned_mapper = umap.AlignedUMAP(**settings_alignment_chosen['umap_kwargs'])\n",
    "    aligned_mapper.embeddings_ = aligned_mapper_embeddings    \n",
    "    \"\"\"\n",
    "    # embedded_datasets = ...LOAD each data_subdict object...\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "embedded_datasets = add_PCA_to_embedded_datasets(embedded_datasets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for key in embedded_datasets.keys():\n",
    "#    embedded_datasets[key]['algos'].pop('pca')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part A: Generate 8 panel figure analagous to Fig. 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gamma_selection = settings_alignment_chosen['subplot_selection']  # [0, 3, 4, 5, 7, 8, 9, 13]\n",
    "fpath = NOTEBOOK_OUTDIR + os.sep + 'picks_nobound_2x4_gammas%d' \\\n",
    "    % (len(settings_alignment_chosen['gamma_list']))\n",
    "\n",
    "plot_args = (\n",
    "    embedded_datasets, \n",
    "    settings_alignment_chosen, \n",
    ")\n",
    "plot_kwargs = {\n",
    "    'annotate_indices': [],\n",
    "    'ext': ['.jpg', '.pdf', '.svg'],\n",
    "    'rasterized': True,\n",
    "    'ss': 6.0,\n",
    "    'edges': False,\n",
    "    'alpha': 1.0,\n",
    "    'black_edges': True,\n",
    "    'skip_cbar': False,\n",
    "}\n",
    "\n",
    "c = 'energies'\n",
    "fpath_spec = fpath + '_energies'\n",
    "plot_aligned_panels_as_grid(*plot_args, fpath_spec, **plot_kwargs, c=c)\n",
    "\n",
    "c = 'nunique'\n",
    "fpath_spec = fpath + '_nunique'\n",
    "plot_aligned_panels_as_grid(*plot_args, fpath_spec, **plot_kwargs, c=c)\n",
    "\n",
    "#c = None\n",
    "#fpath_spec = fpath + '_fixedcolor'\n",
    "#plot_aligned_panels_as_grid(*plot_args, fpath_spec, **plot_kwargs, c=c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part B: Plot distributions of fixed point data (energy, num. unique single cell states) on similar subplot grid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_multi_kwargs = settings_alignment_chosen['alignment_wrapper_kwargs'].copy()\n",
    "X_multi_kwargs.pop('skip_alignment', None)\n",
    "X_multi, num_cells, num_genes = build_X_multi(settings_alignment_chosen['manyruns_paths'], **X_multi_kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fpath = NOTEBOOK_OUTDIR + os.sep + 'picks_nobound_2x4_gammas%d' \\\n",
    "    % (len(settings_alignment_chosen['gamma_list']))\n",
    "        \n",
    "plot_kwargs = {\n",
    "    'nrow': 2, \n",
    "    'ncol': 4, \n",
    "    'ext': ['.jpg', '.pdf'],\n",
    "    'gamma_selection': None\n",
    "}\n",
    "\n",
    "for fpdata_type in ['energies', 'nunique']:\n",
    "    fpath_figure = fpath + '_%sHist' % fpdata_type\n",
    "    plot_args = (\n",
    "        embedded_datasets,\n",
    "        settings_alignment_chosen, \n",
    "        fpdata_type, \n",
    "        fpath_figure\n",
    "    )\n",
    "    plot_fpdata_as_grid(*plot_args, **plot_kwargs, logy=False)\n",
    "    plot_fpdata_as_grid(*plot_args, **plot_kwargs, logy=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part C: Plotly style interactive plot of particular subpanel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from multicell.unsupervised_helper import \\\n",
    "    plotly_express_embedding, generate_control_data, plot_given_multicell, make_dimreduce_object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gamma_selection = settings_alignment_chosen['subplot_selection']\n",
    "gamma_list = settings_alignment_chosen['gamma_list']\n",
    "\n",
    "for idx in gamma_selection:\n",
    "    data_subdict_selected = embedded_datasets[idx]\n",
    "    outdir = NOTEBOOK_OUTDIR + os.sep + 'plotly_idx%d_gamma%.3f' % (idx, gamma_list[idx])\n",
    "    os.makedirs(outdir, exist_ok=True)\n",
    "        \n",
    "    plotly_express_embedding(\n",
    "        data_subdict_selected, \n",
    "        color_by_index=False, \n",
    "        as_landscape=False, \n",
    "        fmod='jupyter', \n",
    "        show=False, \n",
    "        dirpath=outdir, \n",
    "        surf=False, \n",
    "        step=None\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part D: Plot high-demensional state visualization of certain embedded points "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notes from powerpoint figure 5 (From bottom left):\n",
    "- BL: 1395, 3717, 2325, \n",
    "- TL: 2357, 728,\n",
    "- Homogenous + Maze: 636, 3473, \n",
    "- Heterogeneous: 3964, 1135, 1467, 2602"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# constants\n",
    "gamma_list = settings_alignment_chosen['gamma_list']\n",
    "assert gamma_list[13] == 1.0  # assumption for the manually created dicts below\n",
    "subset_orig = [1395, 3717, 2325, 2357, 728, 636, 3473, 3964, 1135, 1467, 2602]\n",
    "\n",
    "# dictionary of points to plot for each data_subdict (i.e. each gamma value UMAP)\n",
    "agg_points_to_plot_v1 = {\n",
    "    i: subset_orig for i in range(len(gamma_list))\n",
    "}\n",
    "\n",
    "agg_points_to_plot_v2 = {\n",
    "    0: [0, 1, 2, 3],\n",
    "    3: [8011, 5709, 8249, 6626, 9345, 9017, 8573, 9229],\n",
    "    4: [9521, 9904, 9691, 5542, 5402, 8322, 7098, 9852, 9844, 8186],\n",
    "    5: [9521, 9904, 9691, 5542, 5402, 8322, 7098, 9852, 9844, 8186],\n",
    "    7: [3886, 1551, 3649, 5492, 3809, 2990, 83, 1787, 8007, 7342, 7882, 4307],\n",
    "    8: [3886, 1551, 3649, 5492, 3809, 2990, 83, 1787, 8007, 7342, 7882, 4307, 9168, 775, 1512, 37],\n",
    "    9: [3886, 1551, 3649, 5492, 3809, 2990, 83, 1787, 8007, 7342, 7882, 4307, 9168, 775, 1512, 37],\n",
    "    13: subset_orig,  # Panel 8: original points\n",
    "}\n",
    "\n",
    "agg_points_to_plot_v3_many = {\n",
    "    13: [2269, 7142,\n",
    "         7350, 7694,\n",
    "         6809, 2, 2947,\n",
    "         8552,         \n",
    "         9713, 369, 52, 2970, 8237, 6789, \n",
    "         6681, 9018, 2722, 2503, 503, 3895, 7831, \n",
    "         9385, 1347, 2458, 5189, 7057,\n",
    "         9742, 5113,\n",
    "         8891, 1908, 2717, 3515, 5112, 5592, 7105, 9447, 9660,\n",
    "         1842, 8609, 4302, 6247, 5649, 224, 5776, 2504, 2978, 5727,\n",
    "         7657,\n",
    "         8134, 3657, 7849,\n",
    "         646, 5607, 3132, 4865, 7076,\n",
    "         5552, 8601, 1828, 2258, 3321,\n",
    "         6467, 2220, 3611, 7414, 5522\n",
    "        ]\n",
    "}\n",
    "\n",
    "\n",
    "agg_points_to_plot_v4_selected = {\n",
    "    13: [2269, 7694, 2,           # maze 'A'\n",
    "         6681, 5189, 5113,        # triple/homog 'B'\n",
    "         2717,                    # \"C\"\n",
    "         2978,                    # \"D\"\n",
    "         7657,                    # B'\n",
    "         6467, 5522               # A'\n",
    "        ]\n",
    "}\n",
    "\n",
    "fmod = ''\n",
    "smod = settings_alignment_chosen['alignment_wrapper_kwargs']['smod_last']\n",
    "\n",
    "agg_points_to_plot = agg_points_to_plot_v4_selected\n",
    "\n",
    "# TODO implement this block -- diff points to plot for each gamma value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from multicell.multicell_replot import plot_tissue_given_agg_idx\n",
    "\n",
    "for gamma_idx in agg_points_to_plot.keys():\n",
    "    gamma_val = gamma_list[gamma_idx]\n",
    "    outdir = NOTEBOOK_OUTDIR + os.sep + 'tissue_g%.3f' % gamma_val\n",
    "    os.makedirs(outdir, exist_ok=True)\n",
    "    data_subdict = embedded_datasets[gamma_idx]\n",
    "    \n",
    "    print('Plotting points for subdict #%d' % gamma_idx)\n",
    "    agg_indices = agg_points_to_plot[gamma_idx]\n",
    "    for agg_idx in agg_indices:\n",
    "        title = 'Point: %d' % agg_idx  # None\n",
    "        plot_tissue_given_agg_idx(\n",
    "            data_subdict, agg_idx, fmod, outdir, \n",
    "            state_int=False, smod_last=smod, title=title)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part E: Plot 'annotated' version of alignedUMAP where some elements are circled"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_orig = [1395, 3717, 2325, 2357, 728, 636, 3473, 3964, 1135, 1467, 2602]\n",
    "fpath_annotate = fpath + '_annotate'\n",
    "ss = 6\n",
    "\n",
    "#agg_points_to_annotate = subset_orig\n",
    "agg_points_to_annotate = agg_points_to_plot_v3_many[13]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_args = (\n",
    "    embedded_datasets, \n",
    "    settings_alignment_chosen, \n",
    "    fpath_annotate\n",
    ")\n",
    "plot_kwargs = {\n",
    "    'ss': ss,\n",
    "    'annotate_indices': agg_points_to_annotate\n",
    "}\n",
    "\n",
    "plot_aligned_panels_as_grid(*plot_args, **plot_kwargs, ext=['.pdf', '.jpg'], rasterized=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part F: Plot single panel umap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gamma_idx = 13\n",
    "gamma_val = settings_alignment_chosen['gamma_list'][gamma_idx]\n",
    "assert gamma_val == 1.0\n",
    "fpath = NOTEBOOK_OUTDIR + os.sep + 'umap_singlepanel_g%.3f' % gamma_val\n",
    "\n",
    "annotate_indices = agg_points_to_plot_v4_selected[13] #[]\n",
    "color_choice = 'nunique' # None, 'energies', 'nunique'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "plot_single_umap_panel_gridspec(\n",
    "    embedded_datasets[gamma_idx], fpath, \n",
    "    ext=['.jpg', '.pdf', '.svg'], \n",
    "    ss=10, alpha=1.0, edges=True,\n",
    "    annotate_indices=annotate_indices, \n",
    "    rasterized=True,\n",
    "    c=color_choice,\n",
    "    skip_cbar=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_single_umap_panel_gridspec(\n",
    "    embedded_datasets[gamma_idx], fpath + '_altA', \n",
    "    ext=['.jpg', '.pdf', '.svg'], \n",
    "    ss=10, alpha=1.0, edges=False,\n",
    "    annotate_indices=annotate_indices, \n",
    "    rasterized=True,\n",
    "    c=color_choice,\n",
    "    skip_cbar=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part G: Plot PCA grid and cumsum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gamma_selection = settings_alignment_chosen['subplot_selection']\n",
    "gamma_list = settings_alignment_chosen['gamma_list']\n",
    "fpath_pca_grid = NOTEBOOK_OUTDIR + os.sep + 'pca_picks_2x4'\n",
    "fpath_pca_cumsum = NOTEBOOK_OUTDIR + os.sep + 'pca_picks_cumsum'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*PCA grid in the style of aligned UMAP grid*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_args = (\n",
    "    embedded_datasets, \n",
    "    settings_alignment_chosen, \n",
    ")\n",
    "plot_kwargs = {\n",
    "    'algo': 'pca',\n",
    "    'annotate_indices': [],\n",
    "    'ext': ['.jpg', '.pdf', '.svg'],\n",
    "    'rasterized': True,\n",
    "    'ss': 6.0,\n",
    "    'edges': False,\n",
    "    'alpha': 1.0,\n",
    "    'black_edges': True,\n",
    "    'skip_cbar': False,\n",
    "}\n",
    "\n",
    "c = 'energies'\n",
    "fpath_spec = fpath_pca_grid + '_energies'\n",
    "plot_aligned_panels_as_grid(*plot_args, fpath_spec, **plot_kwargs, c=c)\n",
    "\n",
    "c = 'nunique'\n",
    "fpath_spec = fpath_pca_grid + '_nunique'\n",
    "plot_aligned_panels_as_grid(*plot_args, fpath_spec, **plot_kwargs, c=c)\n",
    "\n",
    "#c = None\n",
    "#fpath_spec = fpath + '_fixedcolor'\n",
    "#plot_aligned_panels_as_grid(*plot_args, fpath_spec, **plot_kwargs, c=c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*PCA cumsum*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# full: (5, 3.5)\n",
    "# inset with xlim(0,25): (2.5, 2.5)\n",
    "fig = plt.figure(figsize=(5, 3.5), dpi=450)\n",
    "for idx in gamma_selection:\n",
    "    gamma_val = gamma_list[idx]\n",
    "    data_subdict_selected = embedded_datasets[idx]\n",
    "    pca_dict = data_subdict_selected['algos']['pca']\n",
    "    plt.plot(\n",
    "        pca_dict['cumsum_x'], \n",
    "        pca_dict['cumsum_y'], \n",
    "        '--',\n",
    "        label=r'$\\gamma=%.2f$' % gamma_val)\n",
    "\n",
    "plt.axvline(300, linestyle='-.', c='k', zorder=1)    \n",
    "\n",
    "# axis modifications\n",
    "#plt.gca().set_xscale('log')\n",
    "#plt.gca().set_yscale('log')\n",
    "#plt.xlim(0,25)\n",
    "\n",
    "#plt.legend()\n",
    "plt.xlabel('Num. components')\n",
    "plt.ylabel('Cumulative sum of eigenvalues')\n",
    "\n",
    "#plt.gcf().subplots_adjust(bottom=0.25, left=0.25)\n",
    "plt.gcf().subplots_adjust(bottom=0.15, left=0.25)\n",
    "plt.savefig(fpath_pca_cumsum + '.svg')\n",
    "plt.savefig(fpath_pca_cumsum + '.pdf')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part H: Correlation between various features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Correlation: self-energy and interaction energy*\n",
    "- use violin plot? \n",
    "- energies array has 5 columns:\n",
    "  - 0: H_quadratic_form (should match index 1)\n",
    "  - 1: H_multi (which is sum of index (2,3,4)\n",
    "  - 2: H_self\n",
    "  - 3: H_app\n",
    "  - 4: H_pairwise_scaled\n",
    "- interesting: compare term 2 and 4 as $\\gamma$ changes\n",
    "  - can use dual violin plot https://seaborn.pydata.org/generated/seaborn.violinplot.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gamma_idx = 2\n",
    "xx = embedded_datasets[gamma_idx]['nunique']\n",
    "yy = embedded_datasets[gamma_idx]['energies'][:,2]\n",
    "print(xx.shape)\n",
    "print(yy.shape)\n",
    "plt.scatter(xx,yy)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gamma_idx = 2\n",
    "xx = embedded_datasets[gamma_idx]['nunique']\n",
    "yy = embedded_datasets[gamma_idx]['energies'][:,4]\n",
    "print(xx.shape)\n",
    "print(yy.shape)\n",
    "plt.scatter(xx,yy)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Correlation: energy and num unique single cell states*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gamma_idx = 0\n",
    "xx = embedded_datasets[gamma_idx]['nunique']\n",
    "yy = embedded_datasets[gamma_idx]['energies'][:,0]\n",
    "\n",
    "df = pd.DataFrame(\n",
    "    {'nunique': xx.astype('str'),\n",
    "     'energies_0': yy}\n",
    ")\n",
    "\n",
    "ax = sns.violinplot(x=\"nunique\", y=\"energies_0\",\n",
    "                    data=df, palette=\"muted\", split=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#sns.set_theme(style=\"whitegrid\")\n",
    "tips = sns.load_dataset(\"tips\")\n",
    "ax = sns.violinplot(x=\"day\", y=\"total_bill\", data=tips)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(tips)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Correlation: init cond to final state*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bug investigation: nunique>8 for $\\gamma=0$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Notes so far**\n",
    "- look in following dir for the state 7000 with nunique=10\n",
    "I:\\Development\\Repositories\\biomodels\\celltypes\\runs\\multicell_manyruns\\Wrandom0_gamma0.00_10k_periodic_fixedorderV3_p3_M100\\s7000\\states\n",
    "- there are files X_0.npz, ... X_4.npz, X_last.npz, X_secondlast.npz\n",
    "- issue: X_3.npz, X_4.npz are the same and are valid fixed points. why is X_last.npz slightly different? is it a bit flip corruption or a bug? \n",
    "- explanation A: s7000 directory does not correspond to agg_index 7000 of the aggregate file. can reaggregate and reindex (this will make the manual plot need to be redone unless mapping is found)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reaggregate the data\n",
    "from multicell.multicell_manyruns import aggregate_manyruns\n",
    "\n",
    "\"\"\"\n",
    "dirname = 'Wrandom0_gamma0.00_10k_periodic_fixedorderV3_p3_M100'\n",
    "basedir = RUNS_FOLDER + os.sep + 'multicell_manyruns' + os.sep + dirname\n",
    "aggregate_manyruns(\n",
    "    basedir, \n",
    "    agg_subdir='aggregate',\n",
    "    agg_states=True,\n",
    "    agg_energy=True,\n",
    "    agg_plot=False,\n",
    "    only_last=True)\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_data_subdict_local(data_subdict, X, embedding):\n",
    "    \"\"\"\n",
    "    data_subdict: partially complete data_subdict (path and label keys)\n",
    "    X:            raw data 2d arr (num_runs x num_dimensions)\n",
    "    embedding:    embedded data (num_runs x num_lowdim)\n",
    "    \"\"\"\n",
    "    # Reload data_subdict source files\n",
    "    #  note: we transpose raw data, energies for umap/other functions\n",
    "    manyruns_path = data_subdict['path']\n",
    "    agg_dir = manyruns_path + os.sep + 'aggregate'\n",
    "    fpath_state = agg_dir + os.sep + 'X_aggregate_rerun_last.npz'\n",
    "    fpath_energy = agg_dir + os.sep + 'X_energy_last.npz'\n",
    "    fpath_pickle = manyruns_path + os.sep + 'multicell_template.pkl'\n",
    "    X = np.load(fpath_state)['arr_0'].T          \n",
    "    X_energies = np.load(fpath_energy)['arr_0'].T\n",
    "    with open(fpath_pickle, 'rb') as pickle_file:\n",
    "        # unpickling multicell object\n",
    "        multicell_template = pickle.load(pickle_file)  \n",
    "\n",
    "    # Fill out the rest of data_subdict keys        \n",
    "    num_runs, total_spins = X.shape\n",
    "    data_subdict['data'] = X\n",
    "    data_subdict['index'] = list(range(num_runs))\n",
    "    data_subdict['energies'] = X_energies\n",
    "    data_subdict['num_runs'] = num_runs\n",
    "    data_subdict['total_spins'] = total_spins\n",
    "    data_subdict['multicell_template'] = multicell_template\n",
    "    data_subdict['algos'] = {\n",
    "        'umap': \n",
    "            {'reducer': None,\n",
    "             'embedding': embedding}\n",
    "    }\n",
    "\n",
    "    # Add extra analysis data used within the notebook\n",
    "    X_nunique = get_unique_states_over_ensemble(\n",
    "        X, \n",
    "        multicell_template.num_cells, \n",
    "        multicell_template.num_genes\n",
    "    )\n",
    "    data_subdict['nunique'] = X_nunique.T  # note transpose\n",
    "    assert (\n",
    "        data_subdict['energies'].shape[0] == \n",
    "        data_subdict['nunique'].shape[0])\n",
    "\n",
    "    return data_subdict\n",
    "\n",
    "dirname = 'Wrandom0_gamma0.00_10k_periodic_fixedorderV3_p3_M100'\n",
    "data_subdict_rerun0 = {\n",
    "    'label': dirname,\n",
    "    'path': RUNS_FOLDER + os.sep + 'multicell_manyruns' + os.sep + dirname\n",
    "}\n",
    "data_subdict_rerun0 = build_data_subdict_local(\n",
    "    data_subdict_rerun0, None, None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gamma_idx = 0\n",
    "#data_subdict_gamma0 = embedded_datasets[gamma_idx]\n",
    "data_subdict_gamma0 = data_subdict_rerun0\n",
    "xx = data_subdict_gamma0['nunique']\n",
    "yy = data_subdict_gamma0['energies'][:,0]\n",
    "\n",
    "df = pd.DataFrame(\n",
    "    {'nunique': xx.astype('str'),\n",
    "     'energies_0': yy}\n",
    ")\n",
    "\n",
    "print(df[df['nunique'] == '10'])\n",
    "print(df[df['nunique'] == '9'].head())\n",
    "print(df[df['nunique'] == '8'].head())\n",
    "\n",
    "#agg_idx_n9 = [428, 465, 508]\n",
    "#agg_idx_n10 = [7000]\n",
    "\n",
    "agg_idx_n9 = [590, 593, 696]\n",
    "agg_idx_n10 = [7299]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Plot potential bug indices*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "outdir = NOTEBOOK_OUTDIR + os.sep + 'tissue_bugged_rerun'\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "\n",
    "plot_tissue_given_agg_idx(\n",
    "    data_subdict_gamma0, agg_idx_n10[0], '_bug10', outdir, \n",
    "    state_int=True, smod_last=True, title='bug? %d' % agg_idx_n10[0]\n",
    ")\n",
    "\n",
    "for agg in agg_idx_n9:\n",
    "    plot_tissue_given_agg_idx(\n",
    "        data_subdict_gamma0, agg, '_bug9', outdir, \n",
    "        state_int=True, smod_last=True, title='bug? %d' % agg\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "agg = 0\n",
    "sidelength = 10\n",
    "outpath = outdir + os.sep + 'replot_manual_%d' % agg\n",
    "multicell = data_subdict_gamma0['multicell_template']\n",
    "manyruns_path = data_subdict_gamma0['path']\n",
    "\n",
    "print(manyruns_path)\n",
    "\n",
    "#dir(multicell)\n",
    "#print(multicell.graph_state_arr)\n",
    "\n",
    "agg_dir = manyruns_path + os.sep + 'aggregate'\n",
    "fpath_state = agg_dir + os.sep + 'X_aggregate_rerun_last.npz'\n",
    "\n",
    "X = np.load(fpath_state)['arr_0'].T\n",
    "X_state = X[agg, :]\n",
    "X_state = X_state.reshape(multicell.num_cells, multicell.num_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Now check to see if these states are indeed fixed points/minima*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "agg = 0\n",
    "fname = 'X_last.npz'\n",
    "sdir = 's%d' % agg\n",
    "specific_states_dir = manyruns_path + os.sep + sdir + os.sep + 'states'\n",
    "state_path = specific_states_dir + os.sep + fname\n",
    "\n",
    "X_specific = np.load(state_path)['arr_0'].T\n",
    "print(X_specific.shape)\n",
    "#print(X_specific)\n",
    "replot_scatter_dots_TEST(\n",
    "    X_specific.T, sidelength, outpath,\n",
    "    fmod='_bugfmodSpecific_%s' % fname[:-4], \n",
    "    state_int=True, title='bug? %s %d' % (fname, agg))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from multicell.multicell_replot import FIXED_COLOURMAP \n",
    "import matplotlib as mpl\n",
    "\n",
    "def state_to_label(state):\n",
    "    # Idea: assign integer label (0 to 2^N - 1) to the state\n",
    "    # state acts like binary representation of integers\n",
    "    # \"0\" corresponds to all -1\n",
    "    # 2^N - 1 corresponds to all +1\n",
    "    label = 0\n",
    "    bitlist = ((1+np.sign(np.array(state, dtype=int)))/2).astype(int)  # was np.array, now np.sign to handle xi corruption experiments\n",
    "    for bit in bitlist:\n",
    "        label = (label << 1) | bit\n",
    "    return label\n",
    "\n",
    "\n",
    "def lattice_square_loc_to_int(loc, sidelength):\n",
    "    # maps a two-tuple, for the location of a cell on square grid, to a unique integer\n",
    "    # sidelength is sqrt(num_cells), the edge length of the lattice\n",
    "    assert 0 <= loc[0] < sidelength\n",
    "    assert 0 <= loc[1] < sidelength\n",
    "    x, y = loc[0], loc[1]\n",
    "    return x * sidelength + y\n",
    "\n",
    "\n",
    "def lattice_square_int_to_loc(node_idx, sidelength):\n",
    "    # maps node_idx, the unique int rep of a cell location on the grid, to corresponding two-tuple\n",
    "    # sidelength is sqrt(num_cells), the edge length of the lattice\n",
    "    y = node_idx % sidelength              # remainder from the division mod n\n",
    "    x = int((node_idx - y) / sidelength)   # solve for x\n",
    "    return x, y\n",
    "\n",
    "\n",
    "def replot_scatter_dots_TEST(lattice_state, sidelength, outpath,\n",
    "                        fmod='', state_int=False, cmap=None, title=None,\n",
    "                        ext=['.jpg', '.svg'], rasterized=True):\n",
    "    \"\"\"\n",
    "    Full info morphology plot with grid of 9 genes as dots\n",
    "    \"\"\"\n",
    "\n",
    "    if cmap is None:\n",
    "        cmap = FIXED_COLOURMAP\n",
    "\n",
    "    def state_to_colour_and_morphology(state):\n",
    "        \"\"\"\n",
    "        # assign a unique colour to each state based on a colourmap\n",
    "        cellstate_01 = ((cellstate + 1) / 2).astype(int)\n",
    "        cellstate_brief = str(cellstate_01[2]) + str(cellstate_01[5]) + str(cellstate_01[8])\n",
    "\n",
    "        # eight handpicked colours based on combinations of encoded celltypes\n",
    "        color_dict_brief = {\n",
    "            '000': soft_grey_norm,        # grey (all off)\n",
    "            '100': color_A_pos,           # type A - blue\n",
    "            '010': color_B_pos,           # type B - red\n",
    "            '001': color_C_pos,           # type C - yellow\n",
    "            '101': color_AC,              # type A+C - green\n",
    "            '011': color_BC,              # type B+C - orange\n",
    "            '110': color_AB,              # type A+B - purple\n",
    "            '111': color_anchor_white,    # white (all on)\n",
    "        }\n",
    "\n",
    "        unique_colour = color_dict_brief[cellstate_brief]\n",
    "        \"\"\"\n",
    "        genes = [0,1,2, 3,4,5, 6,7,8]\n",
    "        cellstate_brief = state[genes]\n",
    "        #cellstate_brief = [1,1,1, -1, -1, -1, -1, -1, -1]\n",
    "        #cellstate_brief = [-1,-1,-1, 1, 1, 1, -1, -1, -1]\n",
    "        #cellstate_brief = [-1,-1,-1, -1, -1, -1, 1, 1, 1]\n",
    "\n",
    "        label = state_to_label(cellstate_brief)\n",
    "        unique_colour = cmap[label]\n",
    "        return unique_colour[0:3]\n",
    "\n",
    "    n = sidelength\n",
    "    x = np.zeros(n ** 2)\n",
    "    y = np.zeros(n ** 2)\n",
    "    colors = np.zeros((n**2, 3))\n",
    "    for i in range(n):\n",
    "        for j in range(n):\n",
    "            grid_loc_to_idx = lattice_square_loc_to_int((i,j), sidelength)\n",
    "            cellstate = lattice_state[:, grid_loc_to_idx]\n",
    "            colors[grid_loc_to_idx, :] = state_to_colour_and_morphology(cellstate)\n",
    "            x[grid_loc_to_idx] = j\n",
    "            y[grid_loc_to_idx] = n - i\n",
    "\n",
    "    # plot\n",
    "    #fig = plt.figure(figsize=(12, 12))\n",
    "    #fig, ax = plt.subplots(figsize=(12, 12), dpi=100)\n",
    "    #ax = fig.add_axes([0, 0, 1, 1])  # position: left, bottom, width, height\n",
    "    #ax.set_axis_off()\n",
    "    fig, ax = plt.subplots(figsize=(12, 12), dpi=GLOBAL_DPI)\n",
    "    ax.set_axis_off()\n",
    "\n",
    "    def get_cell_mask(gene_idx):\n",
    "        mask = lattice_state[gene_idx, :] == 1\n",
    "        return mask\n",
    "\n",
    "    # plot - detailed settings\n",
    "    assert n in [2, 10, 20]\n",
    "    if n == 10:\n",
    "        box_lw = 2*1.5\n",
    "        eps = 0.25\n",
    "        lw = 2*2\n",
    "        boxsize = 4*1800  # 600, 750, 850, at 990 it forms grey grid\n",
    "        trisize = 4*50  # 225\n",
    "        lw_eps = 0.05\n",
    "        fontsize = 24\n",
    "\n",
    "    elif n == 2:\n",
    "        box_lw = 4*1.5\n",
    "        eps = 0.3\n",
    "        lw = 4*2\n",
    "        boxsize = 95*1800  # 600, 750, 850, at 990 it forms grey grid\n",
    "        trisize = 30*50  # 225\n",
    "        lw_eps = 0.05\n",
    "        fontsize = 24\n",
    "\n",
    "    else:\n",
    "        assert n == 20\n",
    "        box_lw = 1.5\n",
    "        eps = 0.25\n",
    "        lw = 2\n",
    "        boxsize = 1800  # 600, 750, 850, at 990 it forms grey grid\n",
    "        trisize = 50  # 225\n",
    "        lw_eps = 0.05\n",
    "        fontsize = 24\n",
    "\n",
    "    # create gene markers (the three celltype block functionality is no longer used, now 9 dots)\n",
    "    appendage_style = 'o'  # 1\n",
    "    appendage_z = 2\n",
    "    t_series = [0] * 9\n",
    "    for idx in range(9):\n",
    "        t_mod = mpl.markers.MarkerStyle(marker=appendage_style)\n",
    "        t_series[idx] = t_mod\n",
    "\n",
    "    # outer square with alpha (orig 0.4 alpha)\n",
    "    plt.scatter(x, y, marker='s', c=colors, alpha=1.0, s=boxsize,\n",
    "                ec='k', zorder=1, lw=box_lw, rasterized=rasterized)\n",
    "    # gene 0, 1 mask for celltype A: originally up/down appendage\n",
    "    mask0 = get_cell_mask(0)\n",
    "    mask1 = get_cell_mask(1)\n",
    "    mask2 = get_cell_mask(2)\n",
    "    x0, x1, x2 = -eps, 0, +eps\n",
    "    y0, y1, y2 = +eps, +eps, +eps\n",
    "    plt.scatter(x[mask0]+x0, y[mask0]+y0, marker=t_series[0], c='white', alpha=1.0,\n",
    "                s=trisize, ec='k', zorder=appendage_z, linewidths=lw, rasterized=rasterized)\n",
    "    plt.scatter(x[mask1]+x1, y[mask1]+y1, marker=t_series[1], c='white', alpha=1.0,\n",
    "                s=trisize, ec='k', zorder=appendage_z, linewidths=lw, rasterized=rasterized)\n",
    "    plt.scatter(x[mask2]+x2, y[mask2]+y2, marker=t_series[2], c='white', alpha=1.0,\n",
    "                s=trisize, ec='k', zorder=appendage_z, linewidths=lw, rasterized=rasterized)\n",
    "    # gene 3, 4 mask for celltype B: originally left/right appendage\n",
    "    mask3 = get_cell_mask(3)\n",
    "    mask4 = get_cell_mask(4)\n",
    "    mask5 = get_cell_mask(5)\n",
    "    x3, x4, x5 = x0, x1, x2\n",
    "    y3, y4, y5 = 0, 0, 0\n",
    "    plt.scatter(x[mask3]+x3, y[mask3]+y3, marker=t_series[3], c='white', alpha=1.0,\n",
    "                s=trisize, ec='k', zorder=appendage_z, linewidths=lw, rasterized=rasterized)\n",
    "    plt.scatter(x[mask4]+x4, y[mask4]+y4, marker=t_series[4], c='white', alpha=1.0,\n",
    "                s=trisize, ec='k', zorder=appendage_z, linewidths=lw, rasterized=rasterized)\n",
    "    plt.scatter(x[mask5]+x5, y[mask5]+y5, marker=t_series[5], c='white', alpha=1.0,\n",
    "                s=trisize, ec='k', zorder=appendage_z, linewidths=lw, rasterized=rasterized)\n",
    "    # gene 6, 7 mask for celltype C: originally membrane/circle interior\n",
    "    mask6 = get_cell_mask(6)\n",
    "    mask7 = get_cell_mask(7)\n",
    "    mask8 = get_cell_mask(8)\n",
    "    x6, x7, x8 = x0, x1, x2\n",
    "    y6, y7, y8 = -eps, -eps, -eps\n",
    "    plt.scatter(x[mask6]+x6, y[mask6]+y6, marker=t_series[6], c='white', alpha=1.0,\n",
    "                s=trisize, ec='k', zorder=appendage_z, linewidths=lw, rasterized=rasterized)\n",
    "    plt.scatter(x[mask7]+x7, y[mask7]+y7, marker=t_series[7], c='white', alpha=1.0,\n",
    "                s=trisize, ec='k', zorder=appendage_z, linewidths=lw, rasterized=rasterized)\n",
    "    plt.scatter(x[mask8]+x8, y[mask8]+y8, marker=t_series[8], c='white', alpha=1.0,\n",
    "                s=trisize, ec='k', zorder=appendage_z, linewidths=lw, rasterized=rasterized)\n",
    "\n",
    "    if state_int:\n",
    "        num_cells = lattice_state.shape[1]\n",
    "        for k in range(num_cells):\n",
    "            cellstate = lattice_state[:, k]\n",
    "            label = state_to_label(cellstate)\n",
    "            i, j = lattice_square_int_to_loc(k, n)\n",
    "            #plt.gca().text(j, i, label, color='black', ha='center', va='center')\n",
    "            #plt.gca().text(j, i, '(%d, %d)' % (j, i), color='black', ha='center', va='center')            \n",
    "            xc = j\n",
    "            yc = n - i\n",
    "            plt.gca().text(xc, yc, '%s\\n(%d, %d)' % (label, xc, yc), color='black', ha='center', va='center')\n",
    "            #print(k, '(%d, %d)' % (i, j), ' - ', label, 'coord (%d, %d)' % (j, i))\n",
    "            #print(cellstate)\n",
    "\n",
    "    if title is not None:\n",
    "        plt.title(title, fontsize=fontsize)\n",
    "    # draw gridlines\n",
    "    ax = plt.gca()\n",
    "    plt.axis('off')  # no grid can look nice\n",
    "    #ax.grid(which='major', axis='both', linestyle='-', color='k', linewidth=2)\n",
    "\n",
    "    #ax.set_xticks([], [])\n",
    "    #ax.set_yticks([], [])\n",
    "    xticks = np.arange(-.5, n, 1)\n",
    "    yticks = np.arange(-.5, n, 1)\n",
    "    ax.set_xticks(xticks)\n",
    "    ax.set_yticks(yticks)\n",
    "    ax.xaxis.set_ticklabels(['' for _ in xticks])\n",
    "    ax.yaxis.set_ticklabels(['' for _ in yticks])\n",
    "\n",
    "    # this crops border\n",
    "    plt.xlim(-0.5 - lw_eps, n - 0.5 + lw_eps)\n",
    "    plt.ylim( 0.5 - lw_eps, n + 0.5 + lw_eps)\n",
    "    fig.subplots_adjust(left=0, bottom=0, right=1, top=1, wspace=0, hspace=0)  # unsure\n",
    "    plt.tight_layout(pad=1.2, w_pad=0.5, h_pad=1.0)                            # maybe remove line\n",
    "    # save figure\n",
    "    if title is None:\n",
    "        bbox_inches = None\n",
    "    else:\n",
    "        bbox_inches = 'tight'\n",
    "    if isinstance(ext, list):\n",
    "        for ext_str in ext:\n",
    "            assert ext_str[0] == '.'\n",
    "            plt.savefig(outpath + fmod + ext_str, bbox_inches=bbox_inches, dpi=GLOBAL_DPI)\n",
    "    else:\n",
    "        plt.savefig(outpath + fmod + ext, bbox_inches=bbox_inches, dpi=GLOBAL_DPI)\n",
    "    plt.close()\n",
    "    return\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Now check to see if these states are indeed fixed points/minima*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc_state_init = np.array([-1, -1, 1, 1, 1, 1, 1, 1, 1])\n",
    "hh = np.dot(multicell.matrix_J, sc_state_init)\n",
    "\n",
    "print(np.sign(hh))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc_state_init = np.array([1, 1, -1, 1, 1, 1, 1, 1, 1])\n",
    "hh = np.dot(multicell.matrix_J, sc_state_init)\n",
    "\n",
    "print(np.sign(hh))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Testing code for plotting"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Plotting num unique cell states -- special discrete cmap**\n",
    "\n",
    "Consider also using different markerstyles (triangle, square, etc.) for different nunique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.color_palette(\"hls\", 8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.color_palette('pastel')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.color_palette()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.7"
  },
  "pycharm": {
   "stem_cell": {
    "cell_type": "raw",
    "metadata": {
     "collapsed": false
    },
    "source": []
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
